controllers/topology/machinedeployment_controller.go

```diff
@@ -0,0 +1,439 @@
+	}
+	return res, nil
+}
+
+func (r *MachineDeploymentReconciler) isClusterPaused(ctx context.Context, md *clusterv1.MachineDeployment, msList []*clusterv1.MachineSet) (bool, error) {
+	var cluster *clusterv1.Cluster
+	var err error
+
+	// If the MachineDeployment still exists, get the Cluster from the MachineDeployment.
+	if md != nil {
+		// Fetch the corresponding Cluster.
+		cluster, err = util.GetClusterByName(ctx, r.Client, md.ObjectMeta.Namespace, md.Spec.ClusterName)
+		if err != nil {
+			return false, errors.Wrapf(err, "failed to check if Cluster is paused")
+		}
+	}
+
+	// Otherwise, get the Cluster from the first MachineSet, if possible.
+	if len(msList) > 0 {
+		cluster, err = util.GetClusterByName(ctx, r.Client, msList[0].ObjectMeta.Namespace, msList[0].Spec.ClusterName)
+		if err != nil {
+			return false, errors.Wrapf(err, "failed to check if Cluster is paused")
+		}
+	}
+
+	if cluster == nil {
+		// This should never happen as at least one MachineDeployment or MachineSet should
+		// always exist during reconciliation.
+		return false, errors.Errorf("couldn't find a Cluster to check if it is paused")
+	}
+
+	return annotations.IsPaused(cluster, cluster), nil
+}
+
+// calculateTemplatesInUse returns all templates referenced in non-deleting MachineDeployment and MachineSets.
+func calculateTemplatesInUse(md *clusterv1.MachineDeployment, msList []*clusterv1.MachineSet) (map[string]bool, error) {
+	templatesInUse := map[string]bool{}
+
+	// Templates of the MachineSet are still in use if the MachineSet is not in deleting.
+	for _, ms := range msList {
+		if !ms.DeletionTimestamp.IsZero() {
+			continue
+		}
+
+		bootstrapRef := ms.Spec.Template.Spec.Bootstrap.ConfigRef
+		infrastructureRef := &ms.Spec.Template.Spec.InfrastructureRef
+		if err := addRef(templatesInUse, bootstrapRef, infrastructureRef); err != nil {
+			return nil, errors.Wrapf(err, "failed to add templates of %s to templatesInUse", KRef{Obj: ms})
+		}
+	}
+
+	// MachineDeployment has already been deleted or still exists and is in deleting. In both cases there
+	// are no templates referenced in the MachineDeployment which are still in use, so let's return here.
+	if md == nil || !md.DeletionTimestamp.IsZero() {
+		return templatesInUse, nil
+	}
+
+	// Templates of the MachineDeployment are still in use if the MachineDeployment is not in deleting.
+	bootstrapRef := md.Spec.Template.Spec.Bootstrap.ConfigRef
+	infrastructureRef := &md.Spec.Template.Spec.InfrastructureRef
+	if err := addRef(templatesInUse, bootstrapRef, infrastructureRef); err != nil {
+		return nil, errors.Wrapf(err, "failed to add templates of %s to templatesInUse", KRef{Obj: md})
+	}
+	return templatesInUse, nil
+}
+
+// addFinalizers adds finalizers to the MachineDeployment and the MachineSets, if they aren't already set.
+func (r MachineDeploymentReconciler) addFinalizers(ctx context.Context, md *clusterv1.MachineDeployment, msList []*clusterv1.MachineSet) error {
```

Comment:
Shouldn't we create the MachineDeployments already with the finalizer in place?

Reply:
Yup, no problem.

Reply:
Done (will push later)